### PR TITLE
Updates for LLVM10

### PR DESCRIPTION
**Pull request by alexfikl**
The code changes required for LLVM10 seem to be fairly trivial. I'm not knowledgeable, so they might not be quite right (beyond compiling).

I also made some more non-trivial changes to the buildsystem/CI to test out that everything still works. It seems to build and pass tests on the CI and locally, but let me know if it can be improved!

### CODE REVIEW

**Comment by alexfikl** [comment]
By the way, for the CI see
https://travis-ci.com/github/alexfikl/Oclgrind/builds/166541294

File: CMakeLists.txt [hunk 1]
```diff
@@ -57,15 +56,38 @@ if (${LLVM_PACKAGE_VERSION} VERSION_LESS "5.0")
   message(FATAL_ERROR "LLVM version must be >= 5.0")
 endif()
 
+if (NOT "${CMAKE_SYSTEM_NAME}" STREQUAL "Windows")
+  if (${LLVM_PACKAGE_VERSION} VERSION_LESS "10.0")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+  else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
```
Comment (jrprice) [inline CMakeLists.txt:63]: I think we can just bump the minimum requirement to C++14 at this point to keep things simple. We should also do this using CMake's builtin capability to make this is a bit more flexibile. Can you try this please?
```
# Require C++14
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -57,15 +56,38 @@ if (${LLVM_PACKAGE_VERSION} VERSION_LESS "5.0")
...
-  bitreader bitwriter core coroutines coverage instrumentation ipo irreader
-  linker lto mcparser objcarcopts option target)
+list(FIND LLVM_AVAILABLE_LIBS LLVM _LLVM_SHARED_INDEX)
+if (${_LLVM_SHARED_INDEX} GREATER -1)
+  set(LLVM_LIBS LLVM)
+else()
+  llvm_map_components_to_libnames(LLVM_LIBS
+    bitreader bitwriter core coroutines coverage instrumentation ipo irreader
+    linker lto mcparser objcarcopts option target)
+  if (${LLVM_PACKAGE_VERSION} VERSION_GREATER "9.0")
```
Comment (jrprice) [inline CMakeLists.txt:80]: This didn't work for me on a system with `9.0.1`. Can you make it `VERSION_GREATER_EQUAL "10.0"` please?

File: CMakeLists.txt [hunk 3]
```diff
@@ -57,15 +56,38 @@ if (${LLVM_PACKAGE_VERSION} VERSION_LESS "5.0")
...
+      llvm_map_components_to_libnames(LLVM_EXTRA_LIBS frontendopenmp)
+      set(LLVM_LIBS ${LLVM_LIBS} ${LLVM_EXTRA_LIBS})
+  endif()
+endif()
+
+# https://github.com/jpalus/opencl-clang/commit/4d91f7ecdd31147e7e05ad72d0f5ec2a473f5405
```
Comment (jrprice) [inline CMakeLists.txt:86]: Can you make this a link directly to the LLVM bugzilla bug (https://bugs.llvm.org/show_bug.cgi?id=44870) please?

File: CMakeLists.txt (hunk 2 above)
Comment (alexfikl) [reply CMakeLists.txt:80]: `VERSION_GREATER_EQUAL` seems to have been added in [CMake 3.7](https://cmake.org/cmake/help/latest/release/3.7.html). Is it ok to bump the minimum version to that?

Comment (jrprice) [reply CMakeLists.txt:80]: Sure, I think that's OK (released ~4 years ago).

File: CMakeLists.txt (hunk 1 above)
Comment (alexfikl) [reply CMakeLists.txt]: Fixed in latest version.

File: CMakeLists.txt (hunk 2 above)
Comment (alexfikl) [reply CMakeLists.txt]: Went with rearranging the if a bit because mac 10.11 only had CMake 3.6.

File: CMakeLists.txt (hunk 3 above)
Comment (alexfikl) [reply CMakeLists.txt]: Fixed.

**Comment by alexfikl** [comment]
CI after latest changes:
https://travis-ci.com/github/alexfikl/Oclgrind/builds/168830232

**Comment by jrprice** [comment]
Looks good, thanks!

**Comment by jrprice** [comment]
@alexfikl Do you still need LLVM 10 support in Oclgrind now that LLVM 11 has been released? I'm working on adding support for LLVM 11 right now, but supporting older versions has become awkward due to changes in the vector type APIs, so I'm strongly considering dropping support for everything older than 11.

**Comment by alexfikl** [comment]
I'm fine with moving to LLVM 11+ only. Thank you for asking!

**Pull request merged**